### PR TITLE
Align algorithms defined so far

### PR DESCRIPTION
**Pull request by zolkis**
Using the improved platform specific steps developed for clamp(), update the algorithms defined so far: input, constant, clamp, concat, batch norm, and the create activation steps.

This is to figure out the best steps to be also applied to subsequent algorithms.

@huningxin PTAL

Because the main branch was also merged, a lot of commits appear here, but only this one is relevant:
https://github.com/webmachinelearning/webnn/pull/405/commits/8fac5ca49b10e2c5697b6dab871f0c5a25d2a877

***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/webnn/pull/405.html" title="Last updated on Jun 21, 2023, 6:05 PM UTC (963c9a0)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/405/329f41c...zolkis:963c9a0.html" title="Last updated on Jun 21, 2023, 6:05 PM UTC (963c9a0)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1912,8 +1924,12 @@ partial interface MLGraphBuilder {
                     1. If |dim| is equal to |axis|, add to |desc|.{{MLOperandDescriptor/dimensions}}[|axis|] the value of |inputs|[|index|].{{MLOperandDescriptor/dimensions}}[|dim|].
         1. Let |output| be the result of invoking the <a>create MLOperand</a> steps given [=this=] and |desc|.
             1. If that throws an error, re-throw the error and stop.
-        1. Make a request to the underlying platform to create an operator for this method with |inputs| connected as input and |output| connected as output and store a reference to the [=implementation-defined=] platform object to |output|.{{MLOperand/[[operand]]}}.
-            1. If that fails, throw a "{{DataError}}" {{DOMException}} and stop.
+        1. If any of the following sub-steps fail, throw an "{{OperationError}}" {{DOMException}} and stop.
+            1. Make a request to the underlying platform to:
+                1. Create an [=implementation-defined=] platform operator |concat| for this method, given |inputs| and |axis|.
+                1. Register |inputs|.{{MLOperand/[[operand]]}} as an input to |concat|.
```
Comment (huningxin) [inline index.bs:1930]: concat may have multiple inputs.

Comment (huningxin) [reply index.bs:1930]: concatImpl? be consistent with other ops?

File: index.bs [hunk 2]
```diff
@@ -1725,10 +1735,13 @@ partial interface MLGraphBuilder {
...
-    1. Issue a request to the underlying platform to initialize the batch normalization, passing the arguments |input|, |mean|, |variance| and |options| and given |result| to store the results and |options|. Wait for completion.
-        <div class="informalsteps">
-            1. If |options|.activation [=map/exists=], implementations MAY use it to optimize the operation flow.
-        </div>
+    1. If any of the following sub-steps fail, throw an "{{OperationError}}" {{DOMException}} and stop.
+        1. Make a request to the underlying platform to initialize the batch normalization:
+            1. Create an [=implementation-defined=] platform operator |batchNormImpl| for this method, given |input|, |mean|, |variance| and |options|.
+            1. Register |result| as output to |batchNormImpl|.
+                <div class="informalsteps">
+                    1. If |options|.activation [=map/exists=], implementations MAY use it to optimize the operation flow.
```
Comment (huningxin) [inline index.bs:1743]: if activation exists, we should register it to platform operator batchNormImpl

File: index.bs [hunk 3]
```diff
@@ -1063,8 +1063,10 @@ The {{MLActivation}} objects (including the ones passed as input to methods) are
     1. Set |activation|.{{MLActivation/[[builder]]}} to |builder|.
     1. Set |activation|.{{MLActivation/[[name]]}} to |name|.
     1. If |options| is an [=object=], set |activation|.{{MLActivation/[[options]]}} to |options|.
-    1. Make a request to the underlying platform to bind the [=implementation-defined=] platform operator for |name| to |activation|.{{MLActivation/[[operator]]}}.
-        1. If that fails, throw a "{{TypeError}}" and abort these steps.
+    1. If any of the following sub-steps fail, throw an "{{OperationError}}" {{DOMException}} and stop.
+        1. Make a request to the underlying platform to:
+            1. Create an [=implementation-defined=] platform operator |opImpl| defined for |name| to |activation|.{{MLActivation/[[operator]]}}, given |options|.
```
Comment (huningxin) [inline index.bs:1068]: The creation of platform operator may depend on the type (fields) of |options|. It is not clear how to handle the different type of options.

File: index.bs [hunk 4]
```diff
@@ -1583,8 +1585,11 @@ Create a named {{MLOperand}} based on a descriptor, that can be used as an input
     1. Let |operand| be the result of invoking the <a>create MLOperand</a> steps with [=this=] and |descriptor|.
         1. If that throws, re-throw the exception and stop.
     1. Set |operand|.{{MLOperand/[[name]]}} to |name|.
-    1. Make a request to the underlying platform to register |operand| as an input and store a reference to the corresponding [=implementation-defined=] platform object in |operand|.{{MLOperand/[[operand]]}}.
-        1. If that fails, throw an "{{OperationError}}" {{DOMException}} and abort these steps.
+    1. If any of the following sub-steps fail, throw an "{{OperationError}}" {{DOMException}} and stop.
+        1. Make a request to the underlying platform to:
+            1. Create an [=implementation-defined=] platform input operand |operandImpl| as an input given |operand|.
```
Comment (huningxin) [inline index.bs:1590]: The creation of a platform input operand should depend on operand descriptor for data type, dimensions etc.,

File: index.bs [hunk 5]
```diff
@@ -1617,8 +1622,11 @@ Create a constant {{MLOperand}} that can be used in {{MLGraphBuilder}} methods.
     1. Let |operand| be the result of invoking the <a>create MLOperand</a> steps with [=this=] and |descriptor|.
         1. If that throws, re-throw the exception and stop.
     1. Let |bytes| be the result of invoking [[=get a copy of the bytes held by the buffer source=]] given |bufferView|.
-    1. Make a request to the underlying platform to register |operand| as a tensor constant with |bytes| as value and store a reference to the corresponding [=implementation-defined=] object to |operand|.{{MLOperand/[[operand]]}}.
-        1. If that fails, throw an "{{OperationError}}" {{DOMException}} and stop.
+    1. If any of the following sub-steps fail, throw an "{{OperationError}}" {{DOMException}} and stop.
+        1. Make a request to the underlying platform to:
+            1. Create an [=implementation-defined=] platform operand |constant| for this method, given |operand|.
```
Comment (huningxin) [inline index.bs:1627]: ditto.

File: index.bs (hunk 3 above)
Comment (zolkis) [reply index.bs:1068]: Yes, because it depends on the activation. Maybe this step should be moved out to the specific activation steps.

File: index.bs (hunk 4 above)
Comment (zolkis) [reply index.bs:1590]: That is why I pass the |operand|, and in rest is implementation-defined. I wonder should I just pass the operand.[[descriptor]]? That feels like an implementation detail, and the steps look more complete like this. Can we extract some more generic steps for creating input, or can we leave it in this generic form?

File: index.bs (hunk 2 above)
Comment (zolkis) [reply index.bs:1743]: OK, I was pondering that, too.

File: index.bs (hunk 1 above)
Comment (zolkis) [reply index.bs:1930]: Yes, yes. Multiple rounds of alignment went missing this one. :)

**Comment by anssiko** [comment]
(Merging this PR to the `zk-conventions-integration` integration branch for staging purposes.)
